### PR TITLE
Update transposed convolution lowering to consider multiple batch sizes.

### PR DESCRIPTION
**Pull request by akshaybaviskar**
Current transposed convolution lowering for GPUs assumes the batch size to be 1.
Update transposed convolution lowering to consider multiple batch sizes.

### CODE REVIEW

File: lib/Dialect/mhlo/transforms/lhlo_legalize_to_affine.cc [hunk 1]
```diff
@@ -901,10 +903,16 @@ struct ConvOpConverter : public OpRewritePattern<ConvOp> {
...
 
-    // Create loop nest (`M` and `N` loops) to iterate over input and kernel
-    // tiles and perform matrix multiplication. Also mark these loops as
-    // parallel.
     SmallVector<Value> tileIvs;
+    // Create loop nest `B` to perform transposed convolution for every batch.
+    AffineForOp forOp = rewriter.create<AffineForOp>(loc, 0, inputShape[0]);
+    tileIvs.push_back(forOp.getInductionVar());
+    forOp.tile_spaceAttr(rewriter.getUnitAttr());
+    forOp->setAttr("polyblocks.parallel", rewriter.getUnitAttr());
```
Comment (bondhugula) [inline lib/Dialect/mhlo/transforms/lhlo_legalize_to_affine.cc:911]: Please use `kPolyBlocksParallelAttrName` that already exists.

File: tests/lhlo-legalize-to-affine.mlir [hunk 2]
```diff
@@ -646,6 +646,7 @@ func @transpose_convolution(%input: memref<1x506x58x128xf32>, %filter: memref<4x
 // TRANSCONV-GPU-TENSOR-NEXT:    %[[C1013:.*]] = arith.constant 1013 : index
 // TRANSCONV-GPU-TENSOR-NEXT:    %[[C117:.*]] = arith.constant 117 : index
 // TRANSCONV-GPU-TENSOR-NEXT:    %[[C63:.*]] = arith.constant 63 : index
+// TRANSCONV-GPU-TENSOR-NEXT:    affine.for %[[B:.*]] = 0 to 3 {
```
Comment (bondhugula) [inline tests/lhlo-legalize-to-affine.mlir:649]: Please adjust the indent for the loops below.

**Review by bondhugula — APPROVED** [review]
Looks good to me. I see that you've just packed the entire batch and hence the compute nest remains unaffected.

File: lib/Dialect/mhlo/transforms/lhlo_legalize_to_affine.cc (hunk 1 above)
Comment (akshaybaviskar) [reply lib/Dialect/mhlo/transforms/lhlo_legalize_to_affine.cc:911]: `kPolyBlocksParallelAttrName` is from `polyblocks`, but these changes are in `mlir-hlo`, including a header file from polyblocks gives `file not found` error. This error can be resolved. But is it okay to include a file from `polyblocks` inside `mlir-hlo`?

Comment (bondhugula) [reply lib/Dialect/mhlo/transforms/lhlo_legalize_to_affine.cc:911]: Oh, I see. mlir-hlo can't depend on polyblocks. In that case, anyway, don't hardcode it here: instead, define it as a constant on the top. We should move this lowering to polyblocks repo next.

Comment (akshaybaviskar) [reply lib/Dialect/mhlo/transforms/lhlo_legalize_to_affine.cc:912]: Okay, thanks.

**Review by bondhugula — APPROVED** [review]

**Pull request merged**